fdbclient/NativeAPI.actor.cpp

```diff
@@ -3292,33 +3302,68 @@ void Transaction::setupWatches() {
 	}
 }
 
-ACTOR Future<TransactionCommitCostEstimation> estimateCommitCosts(Transaction* self,
-                                                                  CommitTransactionRef* transaction) {
-	state MutationRef* it = transaction->mutations.begin();
-	state MutationRef* end = transaction->mutations.end();
-	state TransactionCommitCostEstimation trCommitCosts;
+ACTOR Future<Optional<ClientTrCommitCostEstimation>> estimateCommitCosts(Transaction* self,
+                                                                         CommitTransactionRef const * transaction) {
+	state ClientTrCommitCostEstimation trCommitCosts;
 	state KeyRange keyRange;
-	for (; it != end; ++it) {
-		if (it->type == MutationRef::Type::SetValue) {
-			trCommitCosts.bytesWrite += it->expectedSize();
-			trCommitCosts.numWrite++;
-		} else if (it->isAtomicOp()) {
-			trCommitCosts.bytesAtomicWrite += it->expectedSize();
-			trCommitCosts.numAtomicWrite++;
-		} else if (it->type == MutationRef::Type::ClearRange) {
-			trCommitCosts.numClear++;
+	state int i = 0;
+
+	for (; i < transaction->mutations.size(); ++i) {
+		auto* it = &transaction->mutations[i];
+
+		if (it->type == MutationRef::Type::SetValue || it->isAtomicOp()) {
+			trCommitCosts.opsCount++;
+			trCommitCosts.writeCosts += getWriteOperationCost(it->expectedSize());
+		}
+		else if (it->type == MutationRef::Type::ClearRange) {
+			trCommitCosts.opsCount++;
 			keyRange = KeyRange(KeyRangeRef(it->param1, it->param2));
 			if (self->options.expensiveClearCostEstimation) {
-				StorageMetrics m = wait(self->getStorageMetrics(keyRange, std::numeric_limits<int>::max()));
-				trCommitCosts.bytesClearEst += m.bytes;
+				StorageMetrics m = wait(self->getStorageMetrics(keyRange, CLIENT_KNOBS->TOO_MANY));
```

Comment:
I agree it's expensive, so this option is default false. But in snowflake's discussion, we still think we can leave the option to user who wants accurate estimation.

Reply:
I'd like to understand this a little better, as a seemingly innocuous option could result in a single client driving significant load, affecting other clients as a result. We should consider the use-cases for having this extra accuracy, and whether we could put in any protections that would still allow that use-case to work while keeping the risk lower (limiting the number of clears we can process, etc). It's also something that we may not have a great audit trail for, and if someone does this and generates significant load as a result, it may be difficult for a cluster operator to recognize what's happening.

Reply:
The related counters have been added to `DatabaseContext` and `ProxyStat`.